### PR TITLE
Telegram

### PR DESCRIPTION
**Pull request by chinying**
Add support for Telegram as a new channel type.

## Solution
_Refer to linked PRs for more details._
- Telegram backend and frontend (#373)
- Telegram workers (#424)
- Telegram webhook handler (#389)

## Deploy Notes

**New environment variables**:
- Backend:
    - `TELEGRAM_BOT_TOKEN` : Test Telegram bot token to use when developing locally
    - `TELEGRAM_WEBHOOK_URL` : Telegram webhook callback URL to configure bots with 
    - `DEFAULT_COUNTRY`: Default two-letter ISO country code to use in libphonenumber-js
- Frontend:
    - `REACT_APP_GUIDE_SMS_CREDENTIALS_URL` : Renamed guide link for Twilio credentials
    - `REACT_APP_GUIDE_TELEGRAM_CREDENTIALS_URL` : Guide link for Telegram credentials.
- Worker:
    - `TELEGRAM_BOT_TOKEN`: Bot token to use in development

**New dependencies**:
- Backend:
    - `telegraf`: Library to access Telegram API ([repo](https://github.com/telegraf/telegraf))
    - `libphonenumber-js` : Library for validation of phone numbers without dependency on Closure ([repo](https://github.com/catamphetamine/libphonenumber-js))
- Workers:
    - `telegraf`: Library to access Telegram API ([repo](https://github.com/telegraf/telegraf))

**Database schema changes**:
- Add `TELEGRAM` to `enum_campaigns_type`

## TODOs
**Pre-merge**
- [x] Review comments on worker SQL
- [x] Setup production Telegram handler lambda function and prepare `travis.yml` changes (#504)
- [x] Review and merge in shared templating module (#493)
- [x] Review and merge in UI/UX tweaks (#494, #505)
- [ ] Create database users for Telegram handler

**Merge**
- [ ] Add `TELEGRAM_WEBHOOK_URL` and `DEFAULT_COUNTRY` to Elastic Beanstalk environments

**Post-merge**
- [ ] Revert 4e4094a. `telegram` should not be kept as a valid `NODE_ENV` in production
- [ ] Use phone number normalisation (#486) in `TwilioClient` 
- [ ] Clean up `travis.yml` and Telegram-only environment variables
- [ ] Migrate Postman Testing Bot's subscribers and credential from `telegram` database into `postman` database

### CODE REVIEW

**Comment by jeantanzj** [comment]
SMS broke
Screenshot 2020-07-02 at 2 45 08 AM
Screenshot 2020-07-02 at 2 43 19 AM

File: worker/src/telegram/resources/sql/enqueue-messages-telegram.sql [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  selected_campaign_id int;
+BEGIN
+  UPDATE job_queue
+  SET status = 'SENDING', updated_at = clock_timestamp()
+  WHERE id = job_id
+    AND status = 'ENQUEUED'
+	RETURNING campaign_id INTO selected_campaign_id;
+
+  WITH messages AS (
+    -- start with this campaign's messages joined with telegram ids
```
Comment (jeantanzj) [inline worker/src/telegram/resources/sql/enqueue-messages-telegram.sql:13]: Why not start with getting the list of subscribers for this campaign's botid? This is presumably a much shorter list than the entirety of telegram_subscribers table. 

Join the messages with this bot's subscribers. 

Then set the error "Bot subscription not found" for the messages without a corresponding subscriber, without regard to the existence of a telegram id. 

Enqueue the messages with corresponding subscribers.

To the admin, does knowing whether someone has a telegram id matter? If they want to subscribe to the bot, they have to get a telegram id anyway.

File: serverless/telegram-handler/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import { getBotTokenFromId, verifyBotIdRegistered } from './credentials'
+import { handleUpdate } from './bot'
+
+const handler = async (event: any): Promise<{ statusCode: number }> => {
+  try {
+    // Parse botId and Telegram update
+    const { botId, update } = parseEvent(event)
+
+    // Verify botId and fetch bot token
+    const sequelize = await sequelizeLoader()
```
Comment (jeantanzj) [inline serverless/telegram-handler/src/index.ts:12]: If you define this outside of the handler, you can reuse the warm connection for multiple invocations of the lambda.

File: serverless/telegram-handler/src/credentials.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+/**
+ * Verifies that the given bot id is registered, otherwise throws an error.
+ */
+export const verifyBotIdRegistered = async (
+  botId: string,
+  sequelize: Sequelize
+): Promise<void> => {
+  const { exists: botIdExists } = await sequelize.query(
+    `SELECT EXISTS (SELECT * FROM credentials WHERE name = :botId);`,
```
Comment (jeantanzj) [inline serverless/telegram-handler/src/credentials.ts:21]: Strange syntax :/

`SELECT 1 AS exists FROM credentials WHERE name = :botId;`
will return you exists = undefined if it doesn't exist, and exists = 1 if it does.

File: backend/src/telegram/models/telegram-op.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  params!: object
+
+  @Column(DataType.STRING)
+  messageId?: string
+
+  @Column(DataType.STRING)
+  errorCode?: string
+
+  @Column({
+    type: DataType.ENUM(...Object.values(MessageStatus)),
```
Comment (jeantanzj) [inline backend/src/telegram/models/telegram-op.ts:47]: No need to change this. Making a note here that the relevant casting was done in log-job-telegram.sql

File: backend/src/core/services/template-client.class.ts [hunk 5]
```diff
@@ -157,8 +163,12 @@ export default class TemplateClient {
...
   template(templateBody: string, params: { [key: string]: string }): string {
     const parsed = this.parseTemplate(templateBody, params)
-    // Remove extra '\' infront of single quotes and backslashes, added by Squirrelly when it escaped the csv
-    const templated = parsed.tokens.join('').replace(/\\([\\'])/g, '$1')
+    // Remove extra '\' infront of single quotes and backslashes, added by Squirrelly when it escaped the csv.
+    // Remove extra '\' infront of \n added by Squirrelly when it escaped the message body.
+    const templated = parsed.tokens
+      .join('')
+      .replace(/\\([\\'])/g, '$1')
+      .replace(/\\n/g, '\n')
```
[the remark this answers is not in the document]
Comment (lamkeewei) [reply backend/src/core/services/template-client.class.ts:171]: Checked for SMS. Doesn't affect as SMS replaces linebreaks with `<br />`.

**Pull request merged**